### PR TITLE
Bugfix/cached viewcells return the same results with different templates when rendering

### PR DESCRIPTION
**Pull request by garethellis36**
See #9531 

The fix was to add the template name to the cache key. However, there is one use-case where the defect is not fixed. If the caller explicitly sets a cache key, then they would always get the same results from `Cell::render()` regardless of the template.

### CODE REVIEW

File: src/View/Cell.php [hunk 1]
```diff
@@ -232,12 +232,13 @@ public function render($template = null)
      * @param string $action The action invoked.
      * @return array The cache configuration.
      */
-    protected function _cacheConfig($action)
+    protected function _cacheConfig($action, $template = null)
```
Comment (dereuromark) [inline src/View/Cell.php]: template needs a doc block line

**Comment by ADmad** [comment]
> If the caller explicitly sets a cache key, then they would always get the same results from Cell::render() regardless of the template.

That's fine, if one sets custom key it's his responsibility to ensure it's uniqueness.

File: src/View/Cell.php [hunk 2]
```diff
@@ -232,12 +232,13 @@ public function render($template = null)
...
     {
         if (empty($this->_cache)) {
             return [];
         }
-        $key = 'cell_' . Inflector::underscore(get_class($this)) . '_' . $action;
+        $template = $template ?: "default";
+        $key = 'cell_' . Inflector::underscore(get_class($this)) . '_' . $action . '_' . $template;
```
Comment (ADmad) [inline src/View/Cell.php]: Using action name as well as template to generate the cache key seems unnecessary. How about just using template name? That would also avoid having to add new argument to the method.

**Comment by garethellis36** [comment]
@ADmad can you not use the same template with multiple actions, just as in the normal view layer? If it works the same way then you would not necessarily expect `echo Cell::action1()->render("my_template")` to output the same as `echo Cell::action2()->render("my_template")` (excuse my pseudo-code).

**Comment by ADmad** [comment]
@garethellis36 Right, you can, I overlooked that fact.

**Comment by ADmad** [comment]
Tests are failing.

File: src/View/Cell.php (hunk 2 above)
Comment (saeideng) [inline src/View/Cell.php]: Probably add name to to end of this $var ,make confilict with other generated name

Comment (saeideng) [inline src/View/Cell.php]: My suggest added hashed key to end of this $vars

Comment (markstory) [inline src/View/Cell.php]: Do people often re-use templates between actions? That might be the only case where having both the action and template in the cache key would be helpful.

Comment (garethellis36) [inline src/View/Cell.php]: The API certainly allows you to do that (see my comment below).

**Comment by garethellis36** [comment]
Tests should now be passing & i've fixed the docblock.

**Comment by markstory** [comment]
Thanks :+1:

**Pull request merged**